server/src/test/java/org/elasticsearch/action/search/KnnSearchSingleNodeTests.java

```diff
@@ -144,12 +147,123 @@ public void testKnnFilterWithRewrite() throws IOException {
+
+        // Because of the boost, vector_2 results should appear first
+        assertNotNull(response.getHits().getAt(0).field("vector_2"));
+    }
+
+    public void testMultiKnnClausesSameDoc() throws IOException {
+        int numShards = 1 + randomInt(3);
+        Settings indexSettings = Settings.builder().put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, numShards).build();
+
+        XContentBuilder builder = XContentFactory.jsonBuilder()
+            .startObject()
+            .startObject("properties")
+            .startObject("vector")
+            .field("type", "dense_vector")
+            .field("dims", VECTOR_DIMENSION)
+            .field("index", true)
+            .field("similarity", "l2_norm")
+            .endObject()
+            .startObject("vector_2")
+            .field("type", "dense_vector")
+            .field("dims", VECTOR_DIMENSION)
+            .field("index", true)
+            .field("similarity", "l2_norm")
+            .endObject()
+            .endObject()
+            .endObject();
+        createIndex("index", indexSettings, builder);
+
+        for (int doc = 0; doc < 10; doc++) {
+            // Make them have hte same vector. This will allow us to test the recall is the same but scores take into account both fields
+            float[] vector = randomVector();
+            client().prepareIndex("index").setSource("vector", vector, "vector_2", vector).get();
+        }
+        client().admin().indices().prepareRefresh("index").get();
+
+        float[] queryVector = randomVector();
+        // Having the same query vector and same docs should mean our KNN scores are linearly combined if the same doc is matched
+        KnnSearchBuilder knnSearch = new KnnSearchBuilder("vector", queryVector, 5, 50);
+        KnnSearchBuilder knnSearch2 = new KnnSearchBuilder("vector_2", queryVector, 5, 50);
+        SearchResponse responseOneKnn = client().prepareSearch("index")
+            .setKnnSearch(List.of(knnSearch))
+            .addFetchField("*")
+            .setSize(10)
+            .get();
+        SearchResponse responseBothKnn = client().prepareSearch("index")
+            .setKnnSearch(List.of(knnSearch, knnSearch2))
+            .addFetchField("*")
+            .setSize(10)
+            .get();
+
+        // The total hits is k matched docs
+        assertHitCount(responseOneKnn, 5);
+        assertHitCount(responseBothKnn, 5);
+        assertEquals(5, responseOneKnn.getHits().getHits().length);
+        assertEquals(5, responseBothKnn.getHits().getHits().length);
+
+        for (int i = 0; i < responseOneKnn.getHits().getHits().length; i++) {
+            SearchHit oneHit = responseOneKnn.getHits().getHits()[i];
+            SearchHit bothHit = responseBothKnn.getHits().getHits()[i];
+            assertThat(bothHit.getId(), equalTo(oneHit.getId()));
+            assertThat(bothHit.getScore(), greaterThan(oneHit.getScore()));
+        }
+    }
+
```

Comment:
is it worth testing aggregations too when multiple vector queries are provided, just to double check?

Reply:
I will add agg checks to both new tests :)